cunumeric/deferred.py

```diff
@@ -483,6 +483,42 @@ def _copy_store(self, store: Any) -> DeferredArray:
         store_copy.copy(store_to_copy, deep=True)
         return cast(DeferredArray, store_copy)
 
+    def _slice_store(
+        self, k: slice, store: Store, dim: int
+    ) -> tuple(slice, Store):
+        start = k.start
+        end = k.stop
+        step = k.step
+        size = store.shape[dim]
+
+        if (
+            (
+                (start is not None)
+                and (start >= size)
+                and (end is None or end >= size)
+            )
+            or ((start is not None) and (end is not None) and (start >= end))
+            or ((start is None) and (end is not None) and (end == 0))
+        ):
+            start = 0
+            end = 0
+            step = 1
+        elif (start is None) and (end is not None) and (end < 0):
```

Comment:
The "clamping" of the `start` and `end` values should be happening before you check for cases where you need to set `start = end = 0`. The way this is written currently causes you to miss some cases:

```
x = cn.ones((5,))
print(x[-6:])  # returns [1.] instead of [1. 1. 1. 1. 1.]
print(x[:-5])  # crashes
print(x[:-6])  # returns [1. 1. 1. 1.] instead of []
print(x[-2:10])  # works, but make sure you test this case as well 
print(x[-2:-1])  # works, but make sure you test this case as well
print(x[-2:1])  # crashes
```
I would suggest restructuring into something like this:
```
if start is not None:
    if start < 0:
        start += size
    start = max(0, min(size, start))
if end is not None:
    if end < 0:
        end += size
    end = max(0, min(size, end))
if (
    (start is not None and start == size)
    or (end is not None and end == 0)
    or (start is not None and end is not None and start >= end)
):
    start = 0
    end = 0
    step = 1
k = slice(start, end, step)
```